driver-async/src/main/com/mongodb/async/client/OperationExecutorImpl.java

```diff
@@ -69,22 +78,28 @@ public void onResult(final ClientSession clientSession, final Throwable t) {
                 if (t != null) {
                     errHandlingCallback.onResult(null, t);
                 } else {
-                    final AsyncReadBinding binding = getReadWriteBinding(readPreference, readConcern, clientSession,
-                            session == null && clientSession != null);
-                    if (session != null && session.hasActiveTransaction() && !binding.getReadPreference().equals(primary())) {
-                        throw new MongoClientException("Read preference in a transaction must be primary");
-                    }
-                    operation.executeAsync(binding, new SingleResultCallback<T>() {
-                        @Override
-                        public void onResult(final T result, final Throwable t) {
-                            try {
-                                labelException(t, session);
-                                errHandlingCallback.onResult(result, t);
-                            } finally {
-                                binding.release();
-                            }
-                        }
-                    });
+                    getReadWriteBinding(readPreference, readConcern, clientSession,
+                            session == null && clientSession != null,
+                            new SingleResultCallback<AsyncReadWriteBinding>() {
+                                @Override
+                                public void onResult(final AsyncReadWriteBinding binding, final Throwable t) {
+                                    if (session != null && session.hasActiveTransaction()
+                                            && !binding.getReadPreference().equals(primary())) {
+                                        throw new MongoClientException("Read preference in a transaction must be primary");
```

Comment:
You can't throw in callback - you must pass the exception up the callback stack.